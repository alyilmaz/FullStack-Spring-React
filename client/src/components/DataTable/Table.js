
const Table = ({children}) =>{

    return(
        <table>
            {children}
        </table>
    )
}

export default Table;