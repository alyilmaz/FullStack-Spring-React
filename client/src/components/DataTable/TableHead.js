
const TableHead = ({order, orderBy, setOrder, setOrderBy, setPage,  columns}) =>{

    const handleRequestSort = (property) => {
        const isAsc = orderBy === property && order === 'asc';
        setOrder(isAsc ? 'desc' : 'asc');
        setOrderBy(property);
        setPage(0);
      };

    return(
        <thead >
        <tr data-testid="row">
        {columns?.map(column => {
            const sortIcon = () => {
                if (column.id === orderBy) {
                    if (order === 'asc') {
                        return <i id = "up" className="arrow up"></i>
                    }
                    return <i id= "down" className="arrow down"></i>
                } else {
                    return 'Ô∏è-'
                }
        }

    return (
      <th  data-testid="table-head"  key={column.id}>
        <span data-testid="label" >{column.label}</span>
        {column.sortable?<span className="sort-button"><button id = {column.id + "-sort"} data-testid="sort-button" onClick={() => handleRequestSort(column.id)}>{sortIcon()}</button></span>:null}
      </th>
    )
  })}
        </tr>
      </thead>
    )
}

export default TableHead;