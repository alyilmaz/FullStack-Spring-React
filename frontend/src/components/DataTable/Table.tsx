type Props = {
    children?: React.ReactNode;
}

const Table = (props:Props) =>{
    return(
        <table>
            {props.children}
        </table>
    )

}

export default Table;