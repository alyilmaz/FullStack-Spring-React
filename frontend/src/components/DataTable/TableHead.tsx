
interface Props{
    order: string,
    orderBy: string,
    setOrder(input:string):void,
    setOrderBy(input:string):void,
    setPage(input:number):void,
    columns: Column[]
}

export type Column = {
    id: string,
    sortable: boolean,
    label: string
}
const TableHead = (props: Props) =>{

    const handleRequestSort = (property:string):void => {
        const isAsc = props.orderBy === property && props.order === 'asc';
        props.setOrder(isAsc ? 'desc' : 'asc');
        props.setOrderBy(property);
        props.setPage(0);
      };

    return(
        <thead >
        <tr data-testid="row">
        {props.columns?.map(column => {
            const sortIcon = () => {
                if (column.id === props.orderBy) {
                    if (props.order === 'asc') {
                        return <i id = "up" className="arrow up"></i>
                    }
                    return <i id= "down" className="arrow down"></i>
                } else {
                    return 'Ô∏è-'
                }
        }

    return (
      <th  data-testid="table-head"  key={column.id}>
        <span data-testid="label" >{column.label}</span>
        {column.sortable?<span className="sort-button"><button id = {column.id + "-sort"} data-testid="sort-button" onClick={() => handleRequestSort(column.id)}>{sortIcon()}</button></span>:null}
      </th>
    )
  })}
        </tr>
      </thead>
    )
}

export default TableHead;