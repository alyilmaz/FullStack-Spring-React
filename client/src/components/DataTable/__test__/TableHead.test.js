/* eslint-disable testing-library/no-node-access */
import React from "react";
import {render, screen} from "@testing-library/react";
import TableHead from "../TableHead";
import Table from "../Table";

test("sorting button must be existed", () =>{
    const columns = [
        { id: "id", sortable: true,   label: "ID" },
        { id: "status",     sortable: false,   label: "Status" }]
    render(<Table><TableHead order = {"asc"} orderBy = {"id"} columns = {columns}/></Table>); 
    expect(screen.getByTestId("sort-button")).toBeTruthy(); 
})

test("sorting button must be non-existed", () =>{
    const columns = [
        { id: "id", sortable: false,   label: "ID" },
        { id: "status",     sortable: false,   label: "Status" }]
    render(<Table><TableHead order = {"asc"} orderBy = {"id"} columns = {columns}/></Table>); 
    expect(screen.queryByTestId("sort-button")).not.toBeTruthy(); 
})

test("head label check", () =>{
    const columns = [
        { id: "id", sortable: true,   label: "ID" },
        { id: "status",     sortable: false,   label: "Status" }]
    render(<Table><TableHead order = {"asc"} orderBy = {"id"} columns = {columns}/></Table>); 
    const spans = screen.queryAllByTestId("label");
    expect(spans[0]).toHaveTextContent(["ID"]); 
    expect(spans[1]).toHaveTextContent(["Status"]); 
})

test("first button is up and second one is -", () =>{
    const columns = [
        { id: "id", sortable: true,   label: "ID" },
        { id: "status",     sortable: true,   label: "Status" }]
    render(<Table><TableHead order = {"asc"} orderBy = {"id"} columns = {columns}/></Table>); 
    const button = screen.queryAllByTestId("sort-button");
    expect(button[0]).toContainElement(document.getElementById("up")); 
    expect(button[1]).toHaveTextContent("-"); 
})


test("first button is - and second one is down", () =>{
    const columns = [
        { id: "id", sortable: true,   label: "ID" },
        { id: "status",     sortable: true,   label: "Status" }]
    render(<Table><TableHead order = {"desc"} orderBy = {"status"} columns = {columns}/></Table>); 
    const button = screen.queryAllByTestId("sort-button");
    expect(button[1]).toContainElement(document.getElementById("down")); 
    expect(button[0]).toHaveTextContent("-"); 
})
