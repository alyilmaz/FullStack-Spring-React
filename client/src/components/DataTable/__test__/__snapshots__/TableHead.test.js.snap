// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create snapshot 1`] = `
<table>
  <thead>
    <tr
      data-testid="row"
    >
      <th
        data-testid="table-head"
      >
        <span
          data-testid="label"
        >
          ID
        </span>
        <span
          className="sort-button"
        >
          <button
            data-testid="sort-button"
            id="id-sort"
            onClick={[Function]}
          >
            Ô∏è-
          </button>
        </span>
      </th>
      <th
        data-testid="table-head"
      >
        <span
          data-testid="label"
        >
          Status
        </span>
        <span
          className="sort-button"
        >
          <button
            data-testid="sort-button"
            id="status-sort"
            onClick={[Function]}
          >
            <i
              className="arrow down"
              id="down"
            />
          </button>
        </span>
      </th>
    </tr>
  </thead>
</table>
`;
